nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -732,8 +734,87 @@ impl StringPrototype {
         Ok(String::from_str(agent, substring).into_value())
     }
 
-    fn split(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
+    /// ### [22.1.3.23 String.prototype.split ( separator, limit )](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.split)
+    fn split(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+
+        // 2. If separator is neither undefined nor null, then
+        let separator = args.get(0);
+
+        if matches!(separator, Value::Undefined | Value::Null) {
+            // TODO: get_method(separator, Symbol.split)
+        }
+
+        // 3. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        let limit = args.get(1);
+        let lim = match limit {
+            // 4. If limit is undefined, lim is 2**32 - 1.
+            Value::Undefined => u32::MAX,
+            // else let lim be ℝ(? ToUint32(limit)).
+            // Note: Fast path for integer parameter.
+            Value::Integer(value) => value.into_i64() as u32,
+            _ => to_uint32(agent, limit)?,
+        };
+
+        // 5. Let R be ? ToString(separator).
+        let r = to_string(agent, separator)?;
+
+        // 6. If lim is zero, return an empty array
+        if lim == 0 {
+            return Ok(create_array_from_list(agent, &[]).into_value());
+        }
+
+        // 7. If separator is undefined, return an array with the whole string
+        if separator.is_undefined() {
+            return Ok(create_array_from_list(agent, &[s.into_value()]).into_value());
+        }
+
+        // 8. Let separatorLength be the length of R.
+        let separator_length = r.len(agent);
+
+        // 9. If separatorLength = 0, the split by characters
+        if separator_length == 0 {
+            let subject = s.as_str(agent);
+            let head = subject.split("");
+
+            let mut results: Vec<Value> = head
+                .enumerate()
+                .skip(1) // Rust's split inserts an empty string in the beginning.
+                .take_while(|(i, _)| *i < lim as usize)
+                .map(|(_, part)| Value::from(SmallString::try_from(part).unwrap()))
+                .collect();
+
+            // Remove the latest empty string if exists
+            results.pop();
```

Comment:
**question:** Are you sure this works currectly? What happens if limit is less than the length of the iterator? Won't this then drop one string that was supposed to be received?

Reply:
This case only happens when the separator is an empty string. You can see the documentation of `split()`:

```rs
/// When the empty string is used as a separator, it separates
/// every character in the string, along with the beginning
/// and end of the string.

let f: Vec<_> = "rust".split("").collect();
assert_eq!(f, &["", "r", "u", "s", "t", ""]);
```

In JavaScript `"rust".split("")` must return `["r", "u", "s", "t"]`, so that's why I remove the first and last elements. The first is removed in  `skip(1)`. The last is removed with `results.pop()`, that, according to the docs:
```
Removes the last element from a vector and returns it, or [`None`] if it is empty.
```

I think this code should work fine (all related tests are passed), but I'm open to change it if you want to implement it differently.

Reply:
Yeah, but now what happens if we do : `"rust".split("", 2)`. This should return `["r", "u"]` but I _think_ it will currently return just `["r"]` because we drop the last one regardless of the limit.

Reply:
That's a good point. And, indeed, it doesn't work. I'm fixing it.

Reply:
Ok, I think it's now fixed:
- The `take_while` was wrong because it didn't take into account the skip() done before.
- Only removes the lastest string if the limit is bigger than the result length.

It's weird that there isn't any test for this use case. I would have detected earlier.